src/omlt/neuralnet/layer.py

```diff
@@ -205,6 +205,141 @@ def _eval(self, x):
+class GNNLayer(DenseLayer):
+    r"""
+    Given the number of nodes :math:`N`, implementing a GNN layer defined by:
+
+    .. math::
+
+        \begin{align*}
+        y_i = \sigma \left(\sum\limits_{j=0}^{F_{in}-1}A_{u,v}w_{ji}x_j+b_i\right) && \forall 0\le i<F_{out}
+        \end{align*}
+
+    where :math:`A` is the adjacency matrix (shared by all GNN layers), :math:`u=\lfloor jN/F_{in}\rfloor` is the node index of :math:`j`-th input feature, :math:`v=\lfloor iN/F_{out}\rfloor` is the node index of :math:`i`-th output feature.
+
+    For example, given a GraphSAGE layer with sum aggregation:
+
+    .. math::
+
+        \begin{align*}
+        \mathbf{y_v} = \sigma\left(\mathbf{w_1}^T\mathbf{x_v}+\mathbf{w_2}^T\sum\limits_{u\in\mathcal N(v)}\mathbf{x_u}+\mathbf{b}\right)
+        \end{align*}
+
+    If the graph structure is fixed, assume that it is a line graph with :math:`N=3` nodes, i.e., the adjacency matrix :math:`A=\begin{pmatrix}1 & 1 & 0\\1 & 1 & 1\\ 0 & 1 & 1\end{pmatrix}`. Then the corresponding GNN layer is defined with parameters:
+
+    .. math::
+
+        \begin{align*}
+        \mathbf{W}=\begin{pmatrix}
+            \mathbf{w_1} & \mathbf{w_2} & \mathbf{0} \\
+            \mathbf{w_2} & \mathbf{w_1} & \mathbf{w_2} \\
+            \mathbf{0} & \mathbf{w_2} & \mathbf{w_1} \\
+        \end{pmatrix},
+        \mathbf{B}=\begin{pmatrix}
+        \mathbf{b}\\\mathbf{b}\\\mathbf{b}
+        \end{pmatrix}
+        \end{align*}
+
+    Otherwise, if the input graph structure is not fixed, all weights and biases should be provided. In this case, the GNN layer is defined with parameters:
+
+    .. math::
+
+        \begin{align*}
+        \mathbf{W}=\begin{pmatrix}
+            \mathbf{w_1} & \mathbf{w_2} & \mathbf{w_2} \\
+            \mathbf{w_2} & \mathbf{w_1} & \mathbf{w_2} \\
+            \mathbf{w_2} & \mathbf{w_2} & \mathbf{w_1} \\
+        \end{pmatrix},
+        \mathbf{B}=\begin{pmatrix}
+        \mathbf{b}\\\mathbf{b}\\\mathbf{b}
+        \end{pmatrix}
+        \end{align*}
+
+    Parameters
+    ----------
+    input_size : tuple
+        the size of the input.
+    output_size : tuple
+        the size of the output.
+    weight : matrix-like
+        the weight matrix.
+    biases : array-like
+        the biases.
+    N : int
+        number of nodes in the graph
+    activation : str or None
+        activation function name
+    input_index_mapper : IndexMapper or None
+        map indexes from this layer index to the input layer index size
+    """
+
+    def __init__(
+        self,
+        input_size,
+        output_size,
+        weights,
+        biases,
+        N,
+        *,
+        activation=None,
+        input_index_mapper=None,
+    ):
+        super().__init__(
+            input_size,
+            output_size,
+            weights=weights,
+            biases=biases,
+            activation=activation,
+            input_index_mapper=input_index_mapper,
+        )
+        assert input_size[-1] % N == 0
```

Comment:
Should this be something like:
`if input_size[-1] % N != 0: `
`  raise ValueError(f"Input size columns {input_size[-1]} not divisible by number of nodes {N}.")` ?

Reply:
Couple of thoughts: 
1) do we have a way of checking against the number of node features? I think if the message states a requirement that isn't what's causing the error, that can be confusing for someone getting the error.
2) since this is an error rather than a warning, I would use "must" rather than "should"